### PR TITLE
Fix push notification crash when server returns non-Map error data

### PR DESCRIPTION
**Pull request by pavelzeman**
#### Summary

Fix `UnexpectedNativeTypeException` crash in Android push notification error handling when the server returns `data` as a String instead of a Map.

#### Ticket Link

[MATTERMOST-MOBILE-ANDROID-AWY6](https://mattermost-mr.sentry.io/issues/6306939631/) — 4,806 events, 574 users. Active since Feb 2025, last seen Mar 19 2026.

#### Release Note

```release-note
Fixed an Android crash in push notification handling when the server returns an unexpected error response format.
```

---

### Root Cause

In `General.kt:18`, the push notification fetch error handler calls `response.getMap("data")` without checking the type first. The crash occurs because the `data` field in the response `ReadableMap` is not always a Map.

**How `data` gets its type:** The `react-native-network-client` library's [`Extensions.kt` (lines 75-86)](https://github.com/mattermost/react-native-network-client/blob/main/android/src/main/java/com/mattermost/networkclient/Extensions.kt#L75-L86) determines the type of `data` dynamically based on response body parsing:

```kotlin
when (val json = JSONTokener(bodyString).nextValue()) {
    is JSONArray -> map.putArray("data", json.toWritableArray())  // Array
    is JSONObject -> map.putMap("data", json.toWritableMap())     // Map ✅ (expected)
    else -> map.putString("data", bodyString)                     // String 💥
}
// catch (_: Exception) → map.putString("data", bodyString)       // String 💥
```

So `data` is set as a String when:
1. The response body is **not valid JSON** (HTML error page, plain text)
2. The response body is valid JSON but **not an object or array** (bare string, number)
3. JSON parsing **throws an exception**

**This is NOT a specific Mattermost server version issue.** The Sentry data shows the crash spread across mobile app versions 2.34.0 through 2.38.0, confirming it's triggered by infrastructure conditions, not a server change.

### When does this happen in practice?

1. **Reverse proxy/load balancer errors** — Nginx, HAProxy, or cloud LB returns an HTML 502/503/504 page instead of a JSON API response. This is the most common scenario for on-premise deployments.

2. **Mattermost server `http.Error()` responses** — Several server code paths return plain-text errors:
   - `channels/app/ratelimit.go:90` → `"limit exceeded"` (429)
   - `channels/app/server.go:811` → `"Use HTTPS"` (400)
   - `channels/app/plugin_requests.go:142,158` → `"Internal Server Error"` (500)
   - `channels/api4/license.go:67` → plain error string (400)
   - `channels/web/static.go` → multiple plain-text error paths (500)

3. **Network-level errors** — Connection resets, TLS failures, or timeouts that produce partial/corrupt response bodies.

### Call Chain

```
Push notification received
→ PushNotificationDataRunnable calls fetch(serverUrl, "/api/v4/...")
→ NetworkClient.request → OkHttp → onResponse
→ Extensions.kt parses response body → data = String (non-JSON body)
→ General.kt resolve() → response.getMap("data") 💥 UnexpectedNativeTypeException
```

### Fix

Check `response.getType("data")` before calling `getMap()`. If the type is `Map`, extract structured error info (status_code, message) as before. If it's any other type (String, Null, Number, Array), include the raw string representation in the error message. This handles all response formats gracefully.

### Changes

- **`android/app/src/main/java/com/mattermost/helpers/push_notification/General.kt`**: Add `ReadableType` check before `getMap("data")` call, with String fallback path.

### Tests

5 unit tests added via standalone test-runner (`test-runner/`) with mocked React Native bridge stubs:
- ✅ Map data type → extracts status_code and message
- ✅ String data type → falls back to getString (HTML/proxy errors)
- ✅ Null data type → handles gracefully
- ✅ Number data type → verifies getMap is never called
- ✅ Null error map → handles gracefully

### Customer Impact (current, before this fix)

**574 users** across mobile app versions 2.34.0–2.38.0 experiencing crashes when their Mattermost server (or a reverse proxy/load balancer in front of it) returns a non-JSON error response during push notification data fetch. This happens silently in the background — users see missed or delayed push notifications with no visible error.

---
_Migrated from #9609 (fork → org branch)_

### CODE REVIEW

File: android/app/src/main/java/com/mattermost/helpers/push_notification/General.kt [hunk 1]
```diff
@@ -15,8 +16,15 @@ internal suspend fun PushNotificationDataRunnable.Companion.fetch(serverUrl: Str
...
-                    val error = response.getMap("data")
-                    cont.resumeWith(Result.failure((IOException("Unexpected code ${error?.getInt("status_code")} ${error?.getString("message")}"))))
+                    // Server may return "data" as a Map (normal error response) or a String
+                    // (e.g. HTML error pages, proxy errors). Guard against type mismatch.
+                    val errorMessage = if (response.getType("data") == ReadableType.Map) {
+                        val error = response.getMap("data")
+                        "Unexpected code ${error?.getInt("status_code")} ${error?.getString("message")}"
+                    } else {
+                        "Unexpected response: ${response.getString("data")}"
+                    }
```
[the remark this answers is not in the document]
Comment (pavelzeman) [reply android/app/src/main/java/com/mattermost/helpers/push_notification/General.kt]: Fixed — replaced with exhaustive `when` expression that handles every `ReadableType` with its type-safe getter. No more `getString` on non-String types. 7 tests pass locally.

File: test-runner/gradlew [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+#
+
+eval "set -- $(
+        printf '%s\n' "$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS" |
+        xargs -n1 |
+        sed ' s~[^-[:alnum:]+,./:=@_]~\\&~g; ' |
+        tr '\n' ' '
+    )" '"$@"'
+
+exec "$JAVACMD" "$@"
```
[the remark this answers is not in the document]
Comment (pavelzeman) [reply test-runner/gradlew:249]: Not applicable — the test-runner is a lightweight JVM project that only runs on Linux CI (GitHub Actions). No Windows CI for this repo. Skipping `gradlew.bat`.

File: test-runner/src/test/kotlin/com/mattermost/helpers/push_notification/GeneralErrorHandlingTest.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+     * This mirrors the exact logic in General.kt lines 19-25.
+     */
+    private fun extractErrorMessage(response: ReadableMap): String {
+        return if (response.getType("data") == ReadableType.Map) {
+            val error = response.getMap("data")
+            "Unexpected code ${error?.getInt("status_code")} ${error?.getString("message")}"
+        } else {
+            "Unexpected response: ${response.getString("data")}"
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pavelzeman) [reply test-runner/src/test/kotlin/com/mattermost/helpers/push_notification/GeneralErrorHandlingTest.kt]: Acknowledged — the error formatting lives inside an anonymous `suspendCoroutine` callback in `General.kt`, so extracting a shared function would mean refactoring the coroutine structure. The test helper now mirrors the exact `when` expression from production. Not ideal but acceptable for this scope.

File: test-runner/src/test/kotlin/com/mattermost/helpers/push_notification/GeneralErrorHandlingTest.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        val msg = extractErrorMessage(response)
+        assertEquals("Unexpected response: null", msg)
+    }
+
+    @Test
+    fun `when data is Number type - does not call getMap`() {
+        val response = mock(ReadableMap::class.java).apply {
+            `when`(getType("data")).thenReturn(ReadableType.Number)
+            `when`(getString("data")).thenReturn("42")
+        }
```
[the remark this answers is not in the document]
Comment (pavelzeman) [reply test-runner/src/test/kotlin/com/mattermost/helpers/push_notification/GeneralErrorHandlingTest.kt]: Fixed — the Number test now uses `getDouble` (matching the production `when` branch) instead of `getString`. Also verifies `getString` is never called, confirming type-safe access.

File: test-runner/gradlew (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavelzeman) [reply test-runner/gradlew:249]: Our CI runs on Linux and none of our engineers are on Windows — `gradlew.bat` is not needed. Resolving this.

File: test-runner/src/test/kotlin/com/mattermost/helpers/push_notification/GeneralErrorHandlingTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (pavelzeman) [reply test-runner/src/test/kotlin/com/mattermost/helpers/push_notification/GeneralErrorHandlingTest.kt]: Addressed — extracted `formatErrorMessage()` into `ErrorFormatting.kt` and the test now calls the production function directly via symlink. The duplicated `extractErrorMessage` helper has been removed.

**Comment by pavelzeman** [comment]
Addressed — added `hasKey` + `getType` checks for both `status_code` and `message` before accessing them, with a fallback to `"Unexpected response: $error"`. Also added comprehensive test coverage for all type branches and edge cases.

File: android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Each [ReadableType] branch uses the type-safe getter to avoid
+ * UnexpectedNativeTypeException from the React Native bridge.
+ */
+fun formatErrorMessage(response: ReadableMap): String {
+    val dataType = if (response.hasKey("data")) response.getType("data") else ReadableType.Null
+    return when (dataType) {
+        ReadableType.Map -> {
+            val error = response.getMap("data")
+            "Unexpected code ${error?.getInt("status_code")} ${error?.getString("message")}"
+        }
```
[the remark this answers is not in the document]
Comment (pavelzeman) [reply android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt:27]: Addressed — added `hasKey` + `getType` checks for both `status_code` and `message` before accessing them, with a fallback to `"Unexpected response: $error"`. Also added comprehensive test coverage for all type branches and edge cases.

File: android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Formats an error message from a server response based on the "data" field's type.
+ *
+ * Each [ReadableType] branch uses the type-safe getter to avoid
+ * UnexpectedNativeTypeException from the React Native bridge.
+ */
+fun formatErrorMessage(response: ReadableMap): String {
+    val dataType = if (response.hasKey("data")) response.getType("data") else ReadableType.Null
+    return when (dataType) {
```
Comment (larkox) [inline android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt:14]: Is this when exhaustive? Do we need a default?

File: test-runner/src/main/kotlin/com/mattermost/helpers/push_notification/ErrorFormatting.kt [hunk 7]
```diff
@@ -0,0 +1 @@
+../../../../../../../../android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt
```
Comment (larkox) [inline test-runner/src/main/kotlin/com/mattermost/helpers/push_notification/ErrorFormatting.kt:1]: This is ugly. Do you know if we have any better alternative?

**Review by larkox — APPROVED** [review]
LGTM. Just a couple of comments.

I am still not very familiar with testing on the native side, so I cannot make much comment on that part. It looks we are adding lots of new stuff, but I understand this is the first test we add here.

File: android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt (hunk 6 above)
Comment (pavelzeman) [reply android/app/src/main/java/com/mattermost/helpers/push_notification/ErrorFormatting.kt:14]: Yes, this is exhaustive. 

All 6 ReadableType values are covered: Map, String, Number, Boolean, Array, Null (lines #28-#32).

My understanding is Kotlin would actually refuse to compile if a branch was missing.

File: test-runner/src/main/kotlin/com/mattermost/helpers/push_notification/ErrorFormatting.kt (hunk 7 above)
Comment (pavelzeman) [reply test-runner/src/main/kotlin/com/mattermost/helpers/push_notification/ErrorFormatting.kt:1]: Agreed, it's not pretty. 

The constraint is that the test-runner is a plain kotlin("jvm") module — it doesn't have Android/RN dependencies, so we can't just add the production source tree via sourceSets without pulling in everything else. A targeted srcDir breaks package paths, and a shared module is overkill for one file. 

The symlink is ugly but it's the most pragmatic option: it targets exactly one file, keeps us testing production code (not a copy), and has zero build complexity. Worth revisiting if we end up sharing more files.
